### PR TITLE
[WIP] Tree generation

### PR DESCRIPTION
**Pull request by caelunshun**
- [x] Implement a `Schematic` type for storing structures and writing parts of them which overlap with a chunk to the chunk
- [ ] Implement tree schematic generators
- [ ] Implement a tree finisher, with tree types/density based on biomes

### CODE REVIEW

**Comment by caelunshun** [comment]
I noticed a severe performance impact from adding this, and noticed a rather peculiar cause when investigating:
Screenshot from 2019-09-14 23-18-51

`bitvec` is for some reason the main overhead for worldgen, which seems somewhat odd...

File: server/src/worldgen/schematic.rs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+pub struct SchematicBuilder {
+    length: usize,
+    height: usize,
+    width: usize,
+    center_x: usize,
+    center_y: usize,
+    center_z: usize,
+}
+
+impl SchematicBuilder {
```
Comment (Unrealrussian) [inline server/src/worldgen/schematic.rs:177]: Wouldn't it be better to collapse with_dimensions into new. I don't see a case where dimensions 0,0,0 would be used.

File: server/src/worldgen/schematic.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_dimensions(mut self, length: usize, height: usize, width: usize) -> Self {
```
Comment (Unrealrussian) [inline server/src/worldgen/schematic.rs:182]: Wouldn't it be better to collapse with_dimensions into new. I don't see a case where dimensions 0,0,0 would be used.

File: server/src/worldgen/schematic.rs (hunk 1 above)
Comment (Unrealrussian) [inline server/src/worldgen/schematic.rs:177]: Also a function where you would fill a region of blocks with a specific blocktype would be a great helper function. It is a repeated pattern in future Schematics, where you would fill a specific part (like walls, floors, treeleaves, etc.) in the schematics.

**Pull request closed without merging**